android/KMEA/app/src/main/java/com/tavultesoft/kmea/KeyboardPickerActivity.java

```diff
@@ -347,65 +302,38 @@ private static void setSelection(int position) {
    */
   private static void switchKeyboard(int position, boolean aPrepareOnly) {
     setSelection(position);
-    int listPosition = (position >= keyboardsList.size()) ? keyboardsList.size()-1 : position;
-    HashMap<String, String> kbInfo = keyboardsList.get(listPosition);
-    String pkgId = kbInfo.get(KMManager.KMKey_PackageID);
-    if (pkgId == null || pkgId.isEmpty()) {
-      pkgId = KMManager.KMDefault_UndefinedPackageID;
-    }
-    String kbId = kbInfo.get(KMManager.KMKey_KeyboardID);
-    String langId = kbInfo.get(KMManager.KMKey_LanguageID);
-    String kbName = kbInfo.get(KMManager.KMKey_KeyboardName);
-    String langName = kbInfo.get(KMManager.KMKey_LanguageName);
-    String kFont = kbInfo.get(KMManager.KMKey_Font);
-    String kOskFont = kbInfo.get(KMManager.KMKey_OskFont);
+    int size = KeyboardController.getInstance().get().size();
+    int listPosition = (position >= size) ? size-1 : position;
+    Keyboard kbInfo = KeyboardController.getInstance().getKeyboardInfo(listPosition);
+    String pkgId = kbInfo.getPackageID();
+    String kbId = kbInfo.getKeyboardID();
+    String langId = kbInfo.getLanguageID();
+    String kbName = kbInfo.getKeyboardName();
+    String langName = kbInfo.getLanguageName();
+    String kFont = kbInfo.getFont();
+    String kOskFont = kbInfo.getOSKFont();
     if(aPrepareOnly)
       KMManager.prepareKeyboardSwitch(pkgId, kbId, langId, kbName);
     else
       KMManager.setKeyboard(pkgId, kbId, langId, kbName, langName, kFont, kOskFont);
```

Comment:
OK, so `prepareKeyboardSwitch` is a newer case, but the other half is definitely more of the same.